### PR TITLE
GH-858: [LS] Added support for the hierarchical document symbols.

### PR DESCRIPTION
**Pull request by kittaakos**
Closes: eclipse/xtext-eclipse#858.

Signed-off-by: Akos Kitta <kittaakos@typefox.io>

### CODE REVIEW

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/IDocumentSymbolService.xtend
Comment (cdietrich) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/IDocumentSymbolService.xtend]: this is one of the apis where i am not sure if i like it. e.g. this could not be used to produce an eclipse outline without using lsp4j

Comment (kittaakos) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/IDocumentSymbolService.xtend]: I am not against modifying, adjusting it, but I would like to understand what is wrong with the `lsp4j` model API? If I understood you correctly, you would like to introduce another API in `xtext-core` that would be used by `xtext-eclipse` and will be transformed into the [`IOutlineNode`](https://github.com/eclipse/xtext-eclipse/blob/master/org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/outline/IOutlineNode.java), right? But `xtext-eclipse` already has `lsp4j` as a dependency, right? Then why to add another API?

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/HierarchicalDocumentSymbolService.xtend
Comment (kittaakos) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/HierarchicalDocumentSymbolService.xtend]: Sure, we can do this but let's move this conversation to [here](
https://github.com/eclipse/xtext-core/pull/862#discussion_r215621247). I think they are related.

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/IDocumentSymbolService.xtend
Comment (cdietrich) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/IDocumentSymbolService.xtend]: no it does not thats the point

Comment (cdietrich) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/IDocumentSymbolService.xtend]: => for api that might be useful toi be reused in eclipse it might help to have a lsp4j independent api e.g. as we do it for content assist

Comment (cdietrich) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/IDocumentSymbolService.xtend]: => we should decide that at per api level if it can be usefully used in more than the lsp4j context. currently we are inconsistent with that cc @tivervac

Comment (cdietrich) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/IDocumentSymbolService.xtend]: => i want at least having this thought over before introducing a new api that can be used from lsp4j only

Comment (kittaakos) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/IDocumentSymbolService.xtend]: Also, please explain to me, what would the user do with Xtext `2.16`? Let assume, there is this new API but they already have the outline implementation in `xtext-eclipse`. What would they do then? I am obviously missing a step... That's why I cannot see the benefit of introducing this new non-`lsp4j` API.

Comment (cdietrich) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/IDocumentSymbolService.xtend]: this is about new projects that target lsp4j and eclipse. right now they would have toi implement the outline twice

Comment (kittaakos) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/IDocumentSymbolService.xtend]: > no it does not thats the point

So `xtext-eclipse` does not have the `lsp4j` dependency; I was not aware of it. Now I see your point. Sure, let's do it, I do not mind introducing a new API here, and the LS can map it to LSP4J and Eclipse can transform it to `IOutlineNode`s.

Comment (cdietrich) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/IDocumentSymbolService.xtend]: see discussion at https://github.com/eclipse/xtext-eclipse/issues/711#issuecomment-419072815

Comment (cdietrich) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/IDocumentSymbolService.xtend]: if we come to the conclusion that it is impossible to come up with a lsp4j independent api that would be ok but then i would prefer to implement it twice

Comment (svenefftinge) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/IDocumentSymbolService.xtend]: In Xtext we always allowed different levels of abstraction. We need this LSP-level. If you want something on top that abstracts over LSP, please build it. But it doesn't need to be part of this PR.

Comment (cdietrich) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/IDocumentSymbolService.xtend]: i can live with that if we dont offer this as api to be used in eclipse

Comment (svenefftinge) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/IDocumentSymbolService.xtend]: Ok

Comment (kittaakos) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/IDocumentSymbolService.xtend]: > i can live with that if we dont offer this as api to be used in eclipse

Should I remove the `@since 2.16` from the source? How should we proceed with this PR?

Although the PR is not super urgent, I would like to get this done if you do not have any objections. Thanks!

One thing that we have to keep in mind before we merge is the LSP4J API. See here https://github.com/eclipse/lsp4j/issues/252.

Comment (cdietrich) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/IDocumentSymbolService.xtend]: we marked provisional api as `@Beta` before. correct @szarnekow @svenefftinge ?

Comment (svenefftinge) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/IDocumentSymbolService.xtend]: Yes, we did that in some places.

**Review by svenefftinge — APPROVED** [review]

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/DocumentExtensions.xtend
Comment (kthoms) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/DocumentExtensions.xtend]: missing `@since 2.16` here

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/LanguageServerImpl.xtend
Comment (kthoms) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/LanguageServerImpl.xtend]: also add `@since` for new protected methods in public API packages

File: org.eclipse.xtext.testing/xtend-gen/org/eclipse/xtext/testing/DocumentSymbolConfiguraiton.java
Comment (kthoms) [inline org.eclipse.xtext.testing/xtend-gen/org/eclipse/xtext/testing/DocumentSymbolConfiguraiton.java]: add `@since`

File: org.eclipse.xtext.testing/xtend-gen/org/eclipse/xtext/testing/DocumentSymbolConfiguraiton.java
Comment (kthoms) [inline org.eclipse.xtext.testing/xtend-gen/org/eclipse/xtext/testing/DocumentSymbolConfiguraiton.java]: add `@since`

**Comment by kittaakos** [comment]
Thanks for the feedback on this PR. One general question, shall we use [`@Beta`](
https://github.com/eclipse/xtext-core/pull/862#discussion_r217642497) or [`@since`](
https://github.com/eclipse/xtext-core/pull/862#discussion_r218820251). Thanks!

**Comment by kthoms** [comment]
I'd say a class once made public available can't be `@Beta`, and adding API methods can't either. And if I see this correctly, `@Beta` is only used on type level in the codebase.

So adding API methods to API classes can't be beta, thus we have to be carefully.

**Comment by cdietrich** [comment]
No Beta says available but not API

File: org.eclipse.xtext.testing/src/org/eclipse/xtext/testing/AbstractLanguageServerTest.xtend
Comment (kthoms) [inline org.eclipse.xtext.testing/src/org/eclipse/xtext/testing/AbstractLanguageServerTest.xtend]: Typo in class name :(

**Comment by kittaakos** [comment]
> So marking as beta can give users a hint on that

👍 

Can we agree that I should get rid of `@since` and use `@Beta` instead?

**Comment by kthoms** [comment]
So API rule incompatibility is blocking this change? If so, please veto @cdietrich .

**Comment by cdietrich** [comment]
No it’s new code. Simply mark it as beta

**Comment by kittaakos** [comment]
> No it’s new code. Simply mark it as beta

This.

Thanks, guys! I apply the changes soon.

**Comment by kittaakos** [comment]
I made the changes.

**Comment by kittaakos** [comment]
If this PR is good to merge, please push the button. Thank you!

**Pull request merged**
